### DISCUSSION TITLE
Transactions for prior years (football)?

### DISCUSSION THREAD

**Discussion by mgwesterman** [cwendt94/espn-api#555] [Q&A]
Very grateful for all the code and knowledge shared here.  It really jump-started me to dig in and start creating some Python code....  Has anyone looked at grabbing not just 'recent' transactions, but transactions for prior years?  I'd like to know, for instance, was a team's success driven by certain acquisitions vs. drafting?  I used _view=mTransactions2_, and getting waiver and free agent moves was pretty straightforward, but trades were not.  There is a **TRADE_ACCEPT** transaction type but it doesn't include the affected players.  It does have a '_relatedTransactionId_' field that seemed an obvious answer, but even when pulling all transactions for the season i could not find hits on those ids.    The info in the view=kona_playercard_ appears to include trades, but backing into the answer by querying every player doesn't seem right.  Not a big deal, and i know it goes beyond what espn-api does, but i went fairly far down the rabbit hole and thought maybe others have tried as well.

**Comment by alexlawless12** [comment]
Was just looking for the exact same thing, but no leads so far. Let me know if you find a way!

**Comment by cwendt94** [comment]
I think this question semi falls under this issue https://github.com/cwendt94/espn-api/issues/546, where recent activity has not been working for some previous years. 

To workaround that issue I am going to look at adding transaction history to the player class. Like you mentioned though this does require kona_playercard API call to get all of that players info. Once added I don't think it would be too to get all the player cards from a teams roster. All you would need to do is get a list of all the `playerIds` you want to get the transaction history from and then can call `league.player_info(playerId=[playerIds])` which will call kona_playercard API with all of those ids and return a list of Player Classes

**Reply by mgwesterman** [reply]
FWIW, i accidentally found when trying to pass a transaction id to kona_playercard (instead of player ids) it returned ALL players' detailed info.  As a hack, i was just going to go through them and find which ones had trades, since that seemed to be the only unreliable part of 'mTransactions2'

**Reply by cwendt94** [reply]
Thanks, this is very helpful! Checking some of the `mTransactions2` calls I see we can add some filters as well for types
`{"transactions":{"filterType":{"value":["WAIVER","WAIVER_ERROR"]}}}`.

Also seems like this discussion is ver similar to some of the new questions in this discussion  https://github.com/cwendt94/espn-api/discussions/170 FYI

**Reply by mgwesterman** [reply]
i think you will want FREEAGENT as a type also

**Reply by JohnSweeney85** [reply]
> kona_playercard API call

Thanks for making this! Can you explain the Kona API call ? is that a different key ?

**Reply by cwendt94** [reply]
Yeah this is basically the call when you click a player to get more details (On ESPN's Website). ESPN makes a different call to get all the specific stats, transaction, and recent news for the player when you open their "player card".

**Comment by cakes89** [comment]
Hi, in case this is useful to anyone, I am a idiot when it comes to programming but i figured this one out in R. Basically just pull ?view=kona_playercard for any given year, instead of ?view=kona_player_info. It will return a json for every player in the universe that year. Players have a Transactions column (?) and within that, an "Items" column. Items typically comes with two elements, one is the move for the player in question and the other is the corresponding move -- e..g when player A is dropped to make room for Player B added, one item corresponds to the player A transaction and the other is for player B. To isolate moves for the primary player -- just filter to instances of the playercard's PlayerId matching the PlayerId element embedded in Items.  I expanded the json in excelt and did that filtering there. This gave me coomprehensive transacction data so i could do stuff like figure out which players have been added the most.

Here's some R code that runs python too, it will generate the json. Sorry if this is obvious and/or ugly. I wrote this all for fantasy baseball in case that makes a difference:

```{python}

######### test playercard pulls#####

import json
import requests

league_id = ###leagueID here###
year = 2025 ###works back as far as 2018###

# scrape uni, save as json
playercard_url = "https://lm-api-reads.fantasy.espn.com/apis/v3/games/FLB/seasons/" + \
        str(year) + "/segments/0/leagues/" + str(league_id) + '?view=kona_playercard'
filters = { "players": { "limit": 10000, "sortDraftRanks": { "sortPriority": 100, "sortAsc": True, "value": "STANDARD" } } }
headers = {'x-fantasy-filter': json.dumps(filters)}
playercard_req = requests.get(playercard_url,
                 headers=headers,
                 cookies={"swid": "{##enter SWID here###}",
                          "espn_s2": "###enter espn_s2 here###"})
playercard = playercard_req.json()

```

```{r}

######### stitch playercard pulls together#####

playercard = py$playercard
exportJSONplayercard <- toJSON(playercard)
write(exportJSONplayercard, "playercard2025.json")

```

**Comment by Quadram13** [comment]
Just opened a [related PR](https://github.com/cwendt94/espn-api/pull/704), so I'm bumping this thread as well.
